### PR TITLE
WIP - MGMT-9599[enhancement]: replace sed for envsubst

### PR DESCRIPTION
**Pull request by fbac**
# Description

Substitute sed in favor of envsubst in templates across the whole codebase:
- sed will still be used in the codebase when it's used to replace strings
- sed will still be used in the codebase when it's used to format an output
- envsubst will be used only to render templates

Fixes [# (issue)](https://issues.redhat.com/browse/MGMT-9599)

## Type of change

Please select the appropriate options:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] This change is a documentation update

## Testing

- [x] Each commit will include a test to ensure nothing breaks
- [x] Test will be incremental, testing one more commit at a time without removing previous changes

## Checklist

- [x] I have performed a self-review of my own code
- [x] If a change is adding a feature, it should require a change to the README.md and the review should catch this.
- [ ] If the change is a fix, it should have an issue. The review should make sure the comments state the issue (not just the number) and it should use the keywords that will close the issue on merge.
- [x] A change should not be merged unless it passes CI or there is a comment/update saying what testing was passed.
- [x] PRs should not be merged unless positively reviewed.

### CODE REVIEW

**Review by iranzo — APPROVED** [review]
looks sane, please get it tested just in case, it should be done really early during deployment process so it shouldn't matter if full e2e doesn't run

**Comment by alknopfler** [comment]
sorry for the delay! What I talked with Juan few month ago is just to create a card to change "all" the substitution ways to use the same (w/e we want sed or subsenv but only 1). To unify this part of the code was the first idea of Parri.

**Comment by fbac** [comment]
In this case I'm using envsubst over sed only in cases where sed is used to render a template (hence envsubst makes sense), not when used to make changes on the fly or some input/output formatting.

I.E:
- In the current PR it makes sense since we're just rendering a template.

vs.

- Examples like the following in ocp-sync.sh, where we perform some change in an already existing file which is not a template:
`sed -i '/^mountopt =.*/d' /etc/containers/storage.conf`

- Or changes on the fly to render a variable or calculate something, like:
`FAILEDPACKAGES=$(cat ${OUTPUTDIR}/mirror-error.log | tr ": " "\n" | grep ${DESTINATION_REGISTRY} | sed "s/${DESTINATION_REGISTRY}//g" | sed "s#^/##g" | sed 's#-#.#g' | sed 's#olm/##g' | sed 's#/#.#g' | sort -u | xargs echo)`

In these last two examples it doesn't make sense to use envsubst as we're not dealing with templates.

TL;DR Basically all the templates in the project will be using envsubst.

**Comment by alknopfler** [comment]
ok, that's why I was thinking... why don't use for all sed??? the idea is to unify it, right? let's move all to sed

**Comment by fbac** [comment]
Files where envsubst is used:
```
contrib/gpu-operator/deploy.sh
desploy-disconnected-registry/deploy.sh
deploy-metallb/deploy.sh
deploy-odf/deploy.sh
deploy-ui/deploy.sh
deploy-hub-local/lab-nfs.sh
shared-utils/verify_preflight.sh
```

The change is quite similar as changing envsubst for sed, so I cancelled the tests to open the issue to discussions for a few days and agree on a common point. Since we all will be developing features in the future seems like a good starting point to decide on a common practice (envsubst vs. sed) so everybody is on the same page.

**Comment by iranzo** [comment]
> Files where envsubst is used:
> 
> ```
> contrib/gpu-operator/deploy.sh
> desploy-disconnected-registry/deploy.sh
> deploy-metallb/deploy.sh
> deploy-odf/deploy.sh
> deploy-ui/deploy.sh
> deploy-hub-local/lab-nfs.sh
> shared-utils/verify_preflight.sh
> ```
> 
> The change is quite similar as changing envsubst for sed, so I cancelled the tests to open the issue to discussions for a few days and agree on a common point. Since we all will be developing features in the future seems like a good starting point to decide on a common practice (envsubst vs. sed) so everybody is on the same page.

Ensubst helps reduce the complexity of variable escaping, etc (I think this was one of the reasons for its implementation, to reduce some issues with vars replacement, multiple vars, etc, like using templates for the files).

As said before, for simple use cases might not be needed to do the replacement if the 'solution' will be more complex than actually it is, as this is something just some styling on the templates itself... it's like converting the actual files to jinja templates just for the sake of doing it in a more elegant way but with increased complexity

**Comment by danielchg** [comment]
I think that use envsubst is better to work with templates, and for me keep the template file outside the script file is good idea as well.

**Comment by fbac** [comment]
So, let's recap:

@alknopfler  mentioned that he doesn't mind as long as we use only one tool, sed or envsubst.
@eifrach  votes for jinja.
@iranzo  votes for envsubst.
@danielchg  votes for envsubst.
In my case I like the unix way, which is using the best tool for each use-case, envsubst for templating and sed for outputs/formatting/etc.

The problem here is: 
- there are use cases which envsubst can't do the things sed does, such as formatting.
- envsubst was developed in order to just render a template by substituting envvars.

An example:

- perfect example of using envsubst in the gpu module, just for rendering a template:
`envsubst <manifests/03-gpu-subscription.yaml | oc --kubeconfig=${EDGE_KUBECONFIG} apply -f -`

- examples of using sed that can't be covered by envsubst (it's not designed for that purpose):

When the file already exists and it's not a template with envvars:
`sed -i '/^mountopt =.*/d' /etc/containers/storage.conf`

When formatting and output or extracting some value:
`FAILEDPACKAGES=$(cat ${OUTPUTDIR}/mirror-error.log | tr ": " "\n" | grep ${DESTINATION_REGISTRY} | sed "s/${DESTINATION_REGISTRY}//g" | sed "s#^/##g" | sed 's#-#.#g' | sed 's#olm/##g' | sed 's#/#.#g' | sort -u | xargs echo)`


So, based on this, if we use envsubst for templating there will still be sed in our code as envsubst can't cover all the use cases.
And in the other way around, if we substitute all existing envsubst for sed, there will only be sed.

So what do we do guys?

**Comment by alknopfler** [comment]
@fbac I totally agree with you...that's why in the past, we used sometimes sed and sometimes envsubst. 
TBH now if we don't have a clear option and the pipe is working fine (this is the functional principle) IMHO we'd not spend more time doing anything unless we change the approach totally to use Jinja or another tool for all cases...

**Comment by iranzo** [comment]
> So, let's recap:
> 
> @alknopfler mentioned that he doesn't mind as long as we use only one tool, sed or envsubst. @eifrach votes for jinja. @iranzo votes for envsubst. @danielchg votes for envsubst. In my case I like the unix way, which is using the best tool for each use-case, envsubst for templating and sed for outputs/formatting/etc.
> 
> The problem here is:
> 
> * there are use cases which envsubst can't do the things sed does, such as formatting.
> * envsubst was developed in order to just render a template by substituting envvars.
> 
> An example:
> 
> * perfect example of using envsubst in the gpu module, just for rendering a template:
>   `envsubst <manifests/03-gpu-subscription.yaml | oc --kubeconfig=${EDGE_KUBECONFIG} apply -f -`
> * examples of using sed that can't be covered by envsubst (it's not designed for that purpose):
> 
> When the file already exists and it's not a template with envvars: `sed -i '/^mountopt =.*/d' /etc/containers/storage.conf`
> 
> When formatting and output or extracting some value: `FAILEDPACKAGES=$(cat ${OUTPUTDIR}/mirror-error.log | tr ": " "\n" | grep ${DESTINATION_REGISTRY} | sed "s/${DESTINATION_REGISTRY}//g" | sed "s#^/##g" | sed 's#-#.#g' | sed 's#olm/##g' | sed 's#/#.#g' | sort -u | xargs echo)`
> 
> So, based on this, if we use envsubst for templating there will still be sed in our code as envsubst can't cover all the use cases. And in the other way around, if we substitute all existing envsubst for sed, there will only be sed.
> 
> So what do we do guys?

envsubst when possible, as said in my comment, having simpler commands will make code more understandable, so it's fine to replace some of them, but if the overengineering to use something else is not worth it, we can work in the same way we had so far.

**Comment by fbac** [comment]
@iranzo yep, that's the unix++ approach.

So... I guess we don't make changes on this issue and we focus our efforts in some other cards? 
And from now on just envsubst when rendering templates and sed for everything else?

**Comment by fbac** [comment]
Well, as dicussed above I'm closing the PR and cleaning branches, etc.

Team decided to move forward with:
- envsubst for template rendering.
- sed for formatting and extracting data, etc.

And jumping now onto real issues and enhancements. :)
